### PR TITLE
Add unittest, backward of array read/write op

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

File: paddle/framework/operator.cc [hunk 1]
```diff
@@ -388,6 +391,20 @@ class RuntimeInferShapeContext : public InferShapeContext {
     }
   }
 
+  VarDesc::VarType GetVarType(const std::string& name) const override {
+    Variable* var = scope_.FindVar(name);
+    if (var->IsType<LoDTensor>()) {
```
Comment (tonyyang-svail) [inline paddle/framework/operator.cc]: this logic should be under `Variable`, something like `return var->GetType()`.

File: paddle/operators/sum_op.cc [hunk 2]
```diff
@@ -39,6 +45,27 @@ class SumOp : public framework::OperatorWithKernel {
...
+          x_vars[0]->Get<framework::SelectedRows>().value().type());
+    } else if (x_vars[0]->IsType<framework::LoDTensorArray>()) {
+      auto& array = x_vars[0]->Get<framework::LoDTensorArray>();
+      for (auto& each : array) {
+        if (each.numel() != 0) {
+          return framework::ToDataType(each.type());
+        }
+      }
+    }
+    PADDLE_THROW("Unexpected branch");
```
Comment (tonyyang-svail) [inline paddle/operators/sum_op.cc]: More helpful message, something like `"Unexpected Type %s", x_vars[0]->GetType()`

File: paddle/operators/sum_op.h [hunk 3]
```diff
@@ -88,6 +89,33 @@ class SumKernel : public framework::OpKernel<T> {
...
+              PADDLE_ENFORCE(out_array[i].lod() == in_array[i].lod());
+              auto in = EigenVector<T>::Flatten(in_array[i]);
+              auto result = EigenVector<T>::Flatten(out_array[i]);
+              result.device(context.GetEigenDevice<Place>()) = result + in;
+            }
+          }
+        }
+      }
+    } else {
+      PADDLE_THROW("Unexpected branch");
```
Comment (tonyyang-svail) [inline paddle/operators/sum_op.h]: More helpful message, something like `"Unexpected Type %s", out_var->GetType()`

Comment (reyoung) [reply paddle/operators/sum_op.h]: Done

File: paddle/operators/sum_op.cc (hunk 2 above)
Comment (reyoung) [reply paddle/operators/sum_op.cc]: Done.

File: paddle/framework/operator.cc (hunk 1 above)
Comment (reyoung) [reply paddle/framework/operator.cc]: Done.

**Pull request merged**